File: sdks/python/meerkat/live.py
"""LiveChannel -- session lifecycle wrapper for the ``live/*`` RPC surface.

Replaces the old ``RealtimeChannel`` that was removed in the live-adapter
MVP. Unlike the old class (which managed a WebSocket connection directly),
``LiveChannel`` delegates transport ownership to the caller: :meth:`open`
returns a dict shaped like ``LiveOpenResult`` containing a transport
bootstrap (URL + token) that the caller connects independently.

The class wraps the ``MeerkatClient.live_*`` methods, binding a session id
at construction time so callers do not need to thread ``session_id`` /
``channel_id`` through every call.

Example::

    from meerkat import MeerkatClient, LiveChannel

    async with MeerkatClient() as client:
        session = await client.create_session("Hello")
        channel = LiveChannel.session(client, session.id)

        open_result = await channel.open()
        # Connect to open_result["transport"]["url"] externally

        await channel.commit_input()
        status = await channel.status()
        print(status)

        await channel.close()
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Literal

if TYPE_CHECKING:
    from .client import MeerkatClient
    from .generated.types import LiveRefreshResult


class LiveChannel:
    """Session lifecycle wrapper for the ``live/*`` RPC methods.

    Construct via the :meth:`session` class method, then call :meth:`open`
    to start the live channel. The returned dict carries a transport
    bootstrap (URL + token) for the caller to connect externally.
    """

    def __init__(
        self,
        client: MeerkatClient,
        session_id: str,
        *,
        turning_mode: Literal["provider_managed", "explicit_commit"] | None = None,
    ) -> None:
        self._client = client
        self._session_id = session_id
        self._turning_mode = turning_mode
        self._channel_id: str | None = None

    @classmethod
    def session(
        cls,
        client: MeerkatClient,
        session_id: str,
        *,
        turning_mode: Literal["provider_managed", "explicit_commit"] | None = None,
    ) -> LiveChannel:
        """Create a ``LiveChannel`` bound to a standalone session."""
        return cls(client, session_id, turning_mode=turning_mode)

    @property
    def client(self) -> MeerkatClient:
        """The underlying ``MeerkatClient``."""
        return self._client

    @property
    def session_id(self) -> str:
        """The session id this channel is bound to."""
        return self._session_id

    @property
    def channel_id(self) -> str | None:
        """The channel id assigned by the server after :meth:`open`."""
        return self._channel_id

    # -- Lifecycle ---------------------------------------------------------

    async def open(self) -> dict[str, Any]:
        """Open the live channel. Calls ``live/open``.

        Stores the returned ``channel_id`` for subsequent operations.
        Returns the full ``LiveOpenResult`` dict (``channel_id``,
        ``transport``, ``capabilities``, ``continuity``).
        """
        result = await self._client.live_open(
            self._session_id, turning_mode=self._turning_mode
        )
        self._channel_id = result.get("channel_id") or result.get("channelId")
        return result

    async def close(self) -> dict[str, Any]:
        """Close the live channel. Calls ``live/close``."""
        return await self._client.live_close(self._require_channel_id())

    # -- Status / Refresh --------------------------------------------------

    async def status(self) -> dict[str, Any]:
        """Get the current status of the live channel. Calls ``live/status``."""
        return await self._client.live_status(self._require_channel_id())

    async def refresh(self) -> LiveRefreshResult:
        """Apply mutable session config to the open channel.

        Calls ``live/refresh``. Does NOT replay canonical history. Identity
        changes (model/provider swaps) require close + reopen.
        """
        return await self._client.live_refresh(self._require_channel_id())

    # -- Input -------------------------------------------------------------

    async def send_input_text(self, text: str) -> dict[str, Any]:
        """Send a text input chunk. Calls ``live/send_input``."""
        return await self._client.live_send_input_text(
            self._require_channel_id(), text
        )

    async def send_input_audio(
        self,
        data_base64: str,
        sample_rate_hz: int,
        channels: int,
    ) -> dict[str, Any]:
        """Send a base64-encoded audio chunk. Calls ``live/send_input``."""
        return await self._client.live_send_input_audio(
            self._require_channel_id(), data_base64, sample_rate_hz, channels
        )

    async def send_input_image(
        self, mime: str, data_base64: str
    ) -> dict[str, Any]:
        """Send a base64-encoded image input chunk. Calls ``live/send_input``."""
        return await self._client.live_send_input_image(
            self._require_channel_id(), mime, data_base64
        )

    async def send_input_video_frame(
        self,
        codec: str,
        data_base64: str,
        timestamp_ms: int,
    ) -> dict[str, Any]:
        """Send a base64-encoded video-frame input chunk. Calls ``live/send_input``."""
        return await self._client.live_send_input_video_frame(
            self._require_channel_id(), codec, data_base64, timestamp_ms
        )

    async def commit_input(
        self, response_modality: str | None = None
    ) -> dict[str, Any]:
        """Commit buffered input on the channel. Calls ``live/commit_input``.

        Args:
            response_modality: Optional modality override for this single
                response (``"text"`` or ``"audio"``). ``None`` keeps the
                channel default.
        """
        return await self._client.live_commit_input(
            self._require_channel_id(), response_modality=response_modality
        )

    async def interrupt(self) -> dict[str, Any]:
        """Interrupt the in-progress assistant turn. Calls ``live/interrupt``."""
        return await self._client.live_interrupt(self._require_channel_id())

    async def truncate(
        self,
        item_id: str,
        content_index: int,
        audio_played_ms: int,
    ) -> dict[str, Any]:
        """Truncate assistant output at the client-tracked playback cursor.

        Calls ``live/truncate``.
        """
        return await self._client.live_truncate(
            self._require_channel_id(), item_id, content_index, audio_played_ms
        )

    # -- Internal ----------------------------------------------------------

    def _require_channel_id(self) -> str:
        if self._channel_id is None:
            raise RuntimeError(
                "LiveChannel has not been opened yet -- call open() first"
            )
        return self._channel_id

File: sdks/typescript/src/live.ts
/**
 * LiveChannel — session lifecycle wrapper for the `live/*` RPC surface.
 *
 * Replaces the old `RealtimeChannel` that was removed in the live-adapter
 * MVP. Unlike the old class (which managed a WebSocket connection directly),
 * `LiveChannel` delegates transport ownership to the caller: `open()`
 * returns a {@link LiveOpenResult} containing a transport bootstrap
 * (URL + token) that the caller connects independently.
 *
 * The class wraps the `MeerkatClient.live*` methods, binding a session id
 * at construction time so callers do not need to thread `session_id` /
 * `channel_id` through every call.
 *
 * @example
 * ```ts
 * import { MeerkatClient, LiveChannel } from "@rkat/sdk";
 *
 * const client = new MeerkatClient();
 * await client.connect();
 *
 * const session = await client.createSession("Hello");
 * const channel = LiveChannel.session(client, session.id);
 *
 * const openResult = await channel.open();
 * // Connect to openResult.transport.url externally (WebSocket, etc.)
 *
 * await channel.commitInput();
 * const status = await channel.status();
 * console.log(status.status);
 *
 * await channel.close();
 * ```
 */

import type {
  LiveChannelParams,
  LiveCommitInputParams,
  LiveOpenParams,
  LiveOpenResult,
  LiveRefreshResult,
  LiveSendInputParams,
  LiveStatusResult,
  LiveTruncateParams,
  LiveInputChunkWire,
  RealtimeTurningMode,
  WireLiveResponseModality,
} from "./generated/types.js";
import type { MeerkatClient } from "./client.js";

export type LiveOpenTransport = NonNullable<LiveOpenParams["transport"]>;

/** Options controlling how the live channel is opened. */
export interface LiveChannelOptions {
  /**
   * Turning mode for the channel. `"provider_managed"` lets the provider's
   * VAD decide when to commit; `"explicit_commit"` requires the caller to
   * invoke {@link LiveChannel.commitInput} explicitly.
   *
   * Text-only input via `commitInput({ responseModality: "text" })` requires
   * `"explicit_commit"` because the OpenAI Realtime API rejects
   * `input_audio_buffer.commit` unless the session was opened with
   * explicit-commit semantics.
   */
  readonly turningMode?: RealtimeTurningMode;
  /** Transport to request from `live/open`. Missing uses the server default. */
  readonly transport?: LiveOpenTransport;
}

export class LiveChannel {
  readonly client: MeerkatClient;
  readonly sessionId: string;
  readonly turningMode?: RealtimeTurningMode;
  readonly transport?: LiveOpenTransport;

  /** The channel id assigned by the server after `open()`. */
  private _channelId: string | undefined;

  private constructor(
    client: MeerkatClient,
    sessionId: string,
    options?: LiveChannelOptions,
  ) {
    this.client = client;
    this.sessionId = sessionId;
    this.turningMode = options?.turningMode;
    this.transport = options?.transport;
  }

  /**
   * Create a `LiveChannel` bound to a standalone session.
   */
  static session(
    client: MeerkatClient,
    sessionId: string,
    options?: LiveChannelOptions,
  ): LiveChannel {
    return new LiveChannel(client, sessionId, options);
  }

  /** The channel id, available after {@link open} completes. */
  get channelId(): string | undefined {
    return this._channelId;
  }

  /**
   * Open the live channel. Calls `live/open` and stores the returned
   * `channel_id` for subsequent operations.
   *
   * The returned {@link LiveOpenResult} contains `transport` (URL + token
   * for the caller to connect externally), `capabilities`, and `continuity`.
   */
  async open(): Promise<LiveOpenResult> {
    const params: LiveOpenParams = { session_id: this.sessionId };
    if (this.turningMode != null) {
      params.turning_mode = this.turningMode;
    }
    if (this.transport != null) {
      params.transport = this.transport;
    }
    const result = await this.client.liveOpen(params);
    this._channelId = result.channel_id;
    return result;
  }

  /**
   * Close the live channel. Calls `live/close`.
   *
   * @throws if the channel has not been opened yet.
   */
  async close(): Promise<void> {
    await this.client.liveClose(this.requireChannelParams());
  }

  /**
   * Commit buffered input on the channel. Calls `live/commit_input`.
   *
   * @param responseModality - Optional modality override for this single
   *   response. Pass `"text"` to suppress audio output on an audio-first
   *   channel, or `"audio"` explicitly. `undefined` keeps the channel default.
   */
  async commitInput(responseModality?: string): Promise<void> {
    const params: LiveCommitInputParams = {
      channel_id: this.requireChannelId(),
    };
    if (responseModality != null) {
      params.response_modality = {
        modality: responseModality,
      } as WireLiveResponseModality;
    }
    await this.client.liveCommitInput(params);
  }

  /**
   * Get the current status of the live channel. Calls `live/status`.
   */
  async status(): Promise<LiveStatusResult> {
    return this.client.liveStatus(this.requireChannelParams());
  }

  /**
   * Apply mutable session config to the open channel. Calls `live/refresh`.
   *
   * Does NOT replay canonical history. Refresh enqueues a single
   * `session.update` carrying the latest projection snapshot's mutable
   * config fields. Identity changes (model/provider swaps) require
   * close + reopen.
   */
  async refresh(): Promise<LiveRefreshResult> {
    return this.client.liveRefresh(this.requireChannelParams());
  }

  /**
   * Send an input chunk to the live channel. Calls `live/send_input`.
   */
  async sendInput(chunk: LiveInputChunkWire): Promise<void> {
    const params: LiveSendInputParams = {
      channel_id: this.requireChannelId(),
      chunk,
    };
    await this.client.liveSendInput(params);
  }

  /**
   * Send a text input chunk to the live channel.
   */
  async sendInputText(text: string): Promise<void> {
    await this.sendInput({ kind: "text", text });
  }

  /**
   * Send a base64-encoded audio chunk to the live channel.
   */
  async sendInputAudio(
    dataBase64: string,
    sampleRateHz: number,
    channels: number,
  ): Promise<void> {
    await this.sendInput({
      kind: "audio",
      data: dataBase64,
      sample_rate_hz: sampleRateHz,
      channels,
    });
  }

  /**
   * Send a base64-encoded image input chunk to the live channel.
   */
  async sendInputImage(mime: string, dataBase64: string): Promise<void> {
    await this.sendInput({ kind: "image", mime, data: dataBase64 });
  }

  /**
   * Send a base64-encoded video-frame input chunk to the live channel.
   */
  async sendInputVideoFrame(
    codec: string,
    dataBase64: string,
    timestampMs: number,
  ): Promise<void> {
    await this.sendInput({
      kind: "video_frame",
      codec,
      data: dataBase64,
      timestamp_ms: timestampMs,
    });
  }

  /**
   * Interrupt the in-progress assistant turn. Calls `live/interrupt`.
   */
  async interrupt(): Promise<void> {
    await this.client.liveInterrupt(this.requireChannelParams());
  }

  /**
   * Truncate assistant output at the client-tracked playback cursor.
   * Calls `live/truncate`.
   */
  async truncate(
    itemId: string,
    contentIndex: number,
    audioPlayedMs: number,
  ): Promise<void> {
    const params: LiveTruncateParams = {
      channel_id: this.requireChannelId(),
      item_id: itemId,
      content_index: contentIndex,
      audio_played_ms: audioPlayedMs,
    };
    await this.client.liveTruncate(params);
  }

  // -- Internal helpers ---------------------------------------------------

  private requireChannelId(): string {
    if (this._channelId == null) {
      throw new Error(
        "LiveChannel has not been opened yet — call open() first",
      );
    }
    return this._channelId;
  }

  private requireChannelParams(): LiveChannelParams {
    return { channel_id: this.requireChannelId() };
  }
}
